pkg/autodiscovery/listeners/kube_services.go

```diff
@@ -225,14 +265,232 @@ func (l *KubeServiceListener) removeService(ksvc *v1.Service) {
+
+	// Subsets
+	if len(first.Subsets) != len(second.Subsets) {
+		return true
+	}
+
+	// Addresses and Ports
+	for i := range first.Subsets {
+		if len(first.Subsets[i].Addresses) != len(second.Subsets[i].Addresses) {
+			return true
+		}
+		if len(first.Subsets[i].Ports) != len(second.Subsets[i].Ports) {
+			return true
+		}
+		// Addresses
+		for j := range first.Subsets[i].Addresses {
+			if first.Subsets[i].Addresses[j].IP != second.Subsets[i].Addresses[j].IP {
+				return true
+			}
+			if first.Subsets[i].Addresses[j].Hostname != second.Subsets[i].Addresses[j].Hostname {
+				return true
+			}
+		}
+		// Ports
+		for j := range first.Subsets[i].Ports {
+			if first.Subsets[i].Ports[j].Port != second.Subsets[i].Ports[j].Port {
+				return true
+			}
+			if first.Subsets[i].Ports[j].Name != second.Subsets[i].Ports[j].Name {
+				return true
+			}
+		}
+	}
+	// No relevant change
+	return false
+}
+
+func (l *KubeServiceListener) createEndpoint(kendpt *v1.Endpoints, firstRun bool) {
+	if kendpt == nil {
+		return
+	}
+	endptId := fmt.Sprintf("%s/%s", kendpt.Namespace, kendpt.Name)
+
+	l.m.Lock()
+	for id, isAnnotated := range l.endpointsAnnotated {
+		if id == endptId && isAnnotated {
+			endpts := processEndpoint(kendpt, firstRun)
+			newEndpts, removedEndpts := diffEndpoints(endpts, l.endpoints[endptId])
+			l.updateEndpoints(newEndpts, removedEndpts, endptId)
+			for _, endpt := range newEndpts {
+				log.Debugf("sending new endpoint: %v", endpt)
+				l.newService <- endpt
+			}
+			for _, endpt := range removedEndpts {
+				log.Debugf("removing endpoint: %v", endpt)
+				l.delService <- endpt
+			}
+		}
+	}
+	l.m.Unlock()
+}
+
+func processEndpoint(kendpt *v1.Endpoints, firstRun bool) []*KubeEndpointService {
+	var newEndpointServices []*KubeEndpointService
+
+	// Hosts and Ports
+	var hosts = make(map[string]string)
+	var ports []ContainerPort
+	for i := range kendpt.Subsets {
+		// Hosts
+		for _, host := range kendpt.Subsets[i].Addresses {
+			hosts[host.IP] = host.Hostname
+		}
+		// Ports
+		for _, port := range kendpt.Subsets[i].Ports {
+			ports = append(ports, ContainerPort{int(port.Port), port.Name})
+		}
+	}
+
+	for ip, host := range hosts {
+		// create a separate AD service per host
+		endpt := &KubeEndpointService{
+			entity:       apiserver.EntityForEndpoints(kendpt.Namespace, kendpt.Name, ip),
+			creationTime: integration.After,
+			hosts:        map[string]string{host: ip},
+			ports:        ports,
+			tags: []string{
+				fmt.Sprintf("kube_endpoint:%s", kendpt.Name),
+				fmt.Sprintf("kube_namespace:%s", kendpt.Namespace),
+			},
+			adID: fmt.Sprintf("%s%s/%s", kubeEndpointIDPrefix, kendpt.Namespace, kendpt.Name),
+		}
+		if firstRun {
+			endpt.creationTime = integration.Before
+		}
+		newEndpointServices = append(newEndpointServices, endpt)
+	}
+	return newEndpointServices
+}
+
+func diffEndpoints(current, old []*KubeEndpointService) ([]*KubeEndpointService, []*KubeEndpointService) {
+	var new []*KubeEndpointService
```

Comment:
nit: Although `new` is not [a reserved keyword](https://golang.org/ref/spec#Keywords), it is [a builtin function](https://golang.org/ref/spec#Allocation), we might want to avoid overriding it